File: app/Adruino/UserCode/PixelProtocolDisplayDriver/message_processor.ts
import { mqtt_connection } from "../../System/ArduinoLibraryClones/mqtt_connection";
import { delay, F, Serial } from "../../System/Simulator/ArduinoGlobals";
import { configuration } from "./DataStructures";
import { search_utils } from "./search_utils";
import { SnakeLights } from "./snake_lights";
import { sprite_sheet } from "./sprite_sheet";

const DeviceControl1 = 0x11;
const STX_StartOfText = 0x02;
const ETB_EndOfBlock = 0x17;
const ETX_EndOfText = 0x03;
const EOT_EndOfTransmission = 0x04;
const C_ControlMessage = 0x43;
const P_PixelMode = 0x50;  
const S_ScrollMode = 0x53;
const T_TextMode = 0x54;
const RS_RecordSeparator = 0x1E;
const Empty = "";

export class message_processor {
    private cfg_: configuration;
    private mqtt_: mqtt_connection;
    private lights_: SnakeLights;

    constructor(c: any, mqtt: mqtt_connection, lights: SnakeLights) {
        this.cfg_ = c;
        this.mqtt_ = mqtt;
        this.lights_ = lights;
    }

    async process(bytes: number[], length: number) {
        if (length == 0)
            return;

        const controlCode = bytes[0];

        if (controlCode !== DeviceControl1)
        {
            Serial.println("Not a device control command.");
            return;
        }

        const controlMode = bytes[1];

        Serial.println("Device Control 1 Message received. Mode is:");
        Serial.println(controlMode);

        switch (controlMode) {
            case P_PixelMode:
                this.processSetPixelsMessage(bytes, length);
                return;
            case C_ControlMessage:
                this.processControlMessage(bytes, length);
                return;
            case T_TextMode:
                await this.processSetTextMessage(bytes, length);
                return;
        }
    }

    processSetPixelsMessage(bytes: number[], length: number) {
        Serial.println(F("processSetPixelsMessage"));

        let inPixelBlock = false;
        let r = 0;
        let g = 0;
        let b = 0;

        for (let i = 0; i < length; i++) {
            let current = bytes[i];

            if (inPixelBlock) {
                const x = current;                
                const y = bytes[++i];

                this.lights_.update_pixel_by_coord(x, y, r, g, b, false);

                const next = bytes[i + 1];

                if (next !== RS_RecordSeparator) {
                    inPixelBlock = false;
                } else {
                    i++;
                }

                continue;
            }

            if (current == EOT_EndOfTransmission) {
                break;
            }

            if (current == ETB_EndOfBlock) {
                inPixelBlock = false;
                continue;
            }

            if (current === STX_StartOfText) {
                r = bytes[++i];
                g = bytes[++i];
                b = bytes[++i];
                inPixelBlock = true;
                continue;
            }
        }

        this.lights_.flush();
    }
    
    processControlMessage(bytes: number[], length: number) {
        Serial.println(F("processControlMessage"));

        const controlMessageType = bytes[3];

        if (controlMessageType == 0) {
            this.lights_.clear();
        }
    }

    async processSetTextMessage(bytes: number[], length: number) {
        const messageModeByte = bytes[2];            
        const scrollSpeedMs = bytes[3];
        const r = bytes[4];
        const g = bytes[5];
        const b = bytes[6];

        const shouldScroll = messageModeByte == 1;

        this.printTextMessage(bytes, length);

        if (shouldScroll) {
            Serial.println(F("Scrolling enabled."));

            const maxWidthOffLeft = sprite_sheet.getTotalWidthOfText(bytes, length) * -1;

            let scrollPosition = this.cfg_.display.width;
            while (scrollPosition > maxWidthOffLeft) {

                this.lights_.clearWithoutFlush();
                this.drawEntireString(bytes, length, scrollPosition, r, g, b);

                scrollPosition--;

                await delay(scrollSpeedMs);
            }

        } else {
            Serial.println(F("Scrolling disabled."));
            this.drawEntireString(bytes, length, 0, r, g, b);
        }
    }

    drawEntireString(bytes: number[], length: number, shiftXpositionBy: number, r: number, g: number, b: number) {

        const textStart = bytes.indexOf(STX_StartOfText) + 1;
        const footerLength = 2;
        const spaceWidth = 1;

        let xPosition = shiftXpositionBy;

        for (let i = textStart; i < length - footerLength; i++) {

            const asciiCode = bytes[i];
            const spriteAndWidth = sprite_sheet.spriteDataFor(asciiCode);

            this.drawSpriteAtPosition(spriteAndWidth.spriteData, spriteAndWidth.width, xPosition, r, g, b);

            xPosition += spriteAndWidth.width + spaceWidth;
        }

        this.lights_.flush();
    }

    drawSpriteAtPosition(sprite: number[], width: number, xOffset: number, r: number, g: number, b: number) {

        let bytePosition = 0;

        for (let y = 0; y < 8; y++) {
            for (let x = 0; x < width; x++) {

                const value = sprite[bytePosition++];
                const alignedX = x + xOffset;

                if (value === 1) {
                    this.lights_.update_pixel_by_coord(alignedX, y, r, g, b, false);
                }
            }
        }
    }

    printTextMessage(bytes: number[], length: number) {
        const textStart = search_utils.indexOf(bytes, length, STX_StartOfText) + 1;
        const textEnd = search_utils.indexOf(bytes, length, ETX_EndOfText);

        Serial.println("processSetTextMessage is:");

        for (let i = textStart; i < textEnd; i++) {
            let character = bytes[i];
            Serial.print(character);
      }

        Serial.println("");
    }
    

}

File: arduino/InteractiveLights/MessageProcessor.cpp
#include "MessageProcessor.h"
#include "SnakeLights.h"
#include "SpriteSheet.h"
#include "SearchUtils.h"

const byte DeviceControl1 = 0x11;
const byte STX_StartOfText = 0x02;
const byte ETB_EndOfBlock = 0x17;
const byte ETX_EndOfText = 0x03;
const byte EOT_EndOfTransmission = 0x04;
const byte C_ControlMessage = 0x43;
const byte P_PixelMode = 0x50;
const byte S_ScrollMode = 0x53;
const byte T_TextMode = 0x54;
const byte RS_RecordSeparator = 0x1E;

String Empty = "";

sprite_sheet sprites;

void message_processor::process(char bytes[], int length)
{
  if (length == 0)
    return;

  auto controlCode = (int)bytes[0];
  if (controlCode != DeviceControl1)
  {
    Serial.println("Not a device control command.");
    return;
  }

  auto controlMode = bytes[1];

  Serial.println("Device Control 1 Message received. Mode is:");
  Serial.println((char)controlMode);

  switch (controlMode)
  {
  case P_PixelMode:
    processSetPixelsMessage(bytes, length);
    return;
  case C_ControlMessage:
    processControlMessage(bytes, length);
    return;
  case T_TextMode:
    processSetTextMessage(bytes, length);
    return;
  }
}

void message_processor::processSetPixelsMessage(const char *bytes, int length)
{
  Serial.println(F("processSetPixelsMessage"));

  auto inPixelBlock = false;
  auto r = 0;
  auto g = 0;
  auto b = 0;

  for (auto i = 0; i < length; i++)
  {
    auto current = (int)bytes[i];

    if (inPixelBlock)
    {
      auto x = current;
      auto y = (int)bytes[++i];

      lights_->update_pixel_by_coord(x, y, r, g, b, false);

      auto next = bytes[i + 1];

      if (next != RS_RecordSeparator)
      {
        inPixelBlock = false;
      }
      else
      {
        i++;
      }

      continue;
    }

    if (current == EOT_EndOfTransmission)
    {
      break;
    }

    if (current == ETB_EndOfBlock)
    {
      inPixelBlock = false;
      continue;
    }

    if (current == STX_StartOfText)
    {
      r = bytes[++i];
      g = bytes[++i];
      b = bytes[++i];
      inPixelBlock = true;
      continue;
    }
  }

  lights_->flush();
}

void message_processor::processControlMessage(const char *bytes, int length)
{
  Serial.println(F("processControlMessage"));

  auto controlMessageType = (int)bytes[3];

  if (controlMessageType == 0)
  {
    lights_->clear();
  }
}

void message_processor::processSetTextMessage(const char *bytes, int length)
{
  auto messageModeByte = (int)bytes[2];
  auto scrollSpeedMs = (int)bytes[3];
  auto r = (int)bytes[4];
  auto g = (int)bytes[5];
  auto b = (int)bytes[6];

  auto shouldScroll = messageModeByte == 1;

  printTextMessage(bytes, length);

  if (shouldScroll)
  {
    Serial.println(F("Scrolling enabled."));

    auto maxWidthOffLeft = sprites.getTotalWidthOfText(bytes, length) * -1;

    auto scrollPosition = cfg_->display.width;
    while (scrollPosition > maxWidthOffLeft)
    {
      lights_->clearWithoutFlush();
      drawEntireString(bytes, length, scrollPosition, r, g, b);

      scrollPosition--;

      delay(scrollSpeedMs);
    }
  }
  else
  {
    Serial.println(F("Scrolling disabled."));
    drawEntireString(bytes, length, 0, r, g, b);
  }
}

void message_processor::drawEntireString(const char *bytes, int length, int shiftXpositionBy, int r, int g, int b)
{
  auto textStart = search_utils::indexOf(bytes, length, STX_StartOfText) + 1;
  auto footerLength = 2;
  auto spaceWidth = 1;

  auto xPosition = shiftXpositionBy;

  for (auto i = textStart; i < length - footerLength; i++)
  {
    auto asciiCode = bytes[i];
    auto spriteAndWidth = sprites.spriteDataFor(asciiCode);

    drawSpriteAtPosition(spriteAndWidth.spriteData, spriteAndWidth.width, xPosition, r, g, b);

    xPosition += spriteAndWidth.width + spaceWidth;
  }

  lights_->flush();
}

void message_processor::drawSpriteAtPosition(int *sprite, int width, int xOffset, int r, int g, int b)
{
  auto bytePosition = 0;

  for (auto y = 0; y < 8; y++)
  {
    for (auto x = 0; x < width; x++)
    {
      auto value = sprite[bytePosition++];
      auto alignedX = x + xOffset;

      if (value == 1)
      {
        lights_->update_pixel_by_coord(alignedX, y, r, g, b, false);
      }
    }
  }
}

void message_processor::printTextMessage(const char *bytes, int length)
{
  auto textStart = search_utils::indexOf(bytes, length, STX_StartOfText) + 1;
  auto textEnd = search_utils::indexOf(bytes, length, ETX_EndOfText);

  Serial.println("processSetTextMessage is:");

  for (auto i = textStart; i < textEnd; i++)
  {
    char character = bytes[i];
    Serial.print(character);
  }

  Serial.println("");
}
